### PR TITLE
APPSERV-124 Adds page synchronisation

### PR DESCRIPTION
**Pull request by jbee**
### Summary
Adds server side page configurations with manual and semi-automatic synchronisation options that based on the user role behaves differently. For this purpose roles were added.

* adds a role selection setting and dialog
* adds a page synchronisation buttons in settings (manual) and dialog (semi-automatic)
* adds page config API to webapp and the client controller
* adds new components for role and page dialogues
* adds a formatting for date and time values
* fixes automatic selection of unit combinations when formatting values with time units

Further details can be found in https://github.com/payara/Payara/pull/4666

The API changes required were extracted to a separate PR #14 which needs to be merged first.
Its second last commit should be tagged and released as 1.1.

### Testing 
Must be performed in combination with https://github.com/payara/Payara/pull/4666. See that PR for detailed testing instructions.

### CODE REVIEW

File: webapp/src/main/webapp/js/mc-view.js [hunk 1]
```diff
@@ -682,6 +707,19 @@ MonitoringConsole.View = (function() {
...
     }
 
+    function createRoleSelectionModel(onExit) {
+        return { id: 'RoleSelector', onChange: role => {
+                MonitoringConsole.Model.Role.set(role);
+                $('#RoleSelector').hide();
+                if (onExit !== undefined)
+                    onExit();
+            }, items: [
+            { name: 'guest', label: 'Guest' , description: 'Automatically uses latest server page configuration. Existing local changes are overridden. Local changes during the sesstion do not affect the remote configuration.'},
```
Comment (Pandrex247) [inline webapp/src/main/webapp/js/mc-view.js]:
```suggestion
            { name: 'guest', label: 'Guest' , description: 'Automatically uses latest server page configuration. Existing local changes are overridden. Local changes during the session do not affect the remote configuration.'},
```

File: webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResource.java [hunk 2]
```diff
@@ -130,6 +138,83 @@ private static Series seriesOrNull(String series) {
...
+     */
+    @GET
+    @Path("/pages/data/")
+    public JsonObject getPageData() {
+        JsonObjectBuilder obj = Json.createObjectBuilder();
+        for (String name : pageConfig.listPages()) {
+            String page = pageConfig.getPage(name);
+            if (page != null && !page.isEmpty()) {
+                try (JsonParser parser = Json.createParser(new StringReader(page))) {
+                    if (parser.hasNext()) {
```
Comment (Pandrex247) [inline webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResource.java:161]: Just double-checking that this isn't meant to be a `while` and should only get one thing?

File: webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResource.java [hunk 3]
```diff
@@ -130,6 +138,83 @@ private static Series seriesOrNull(String series) {
...
+     */
+    @GET
+    @Path("/pages/data/{name}/")
+    public String getPageData(@PathParam("name") String name) {
+        return pageConfig.getPage(name);
+    }
+
+    /**
+     * Updates the configuration of the provided page.
+     * If JSON given for the page is null, empty or the empty JSON object the page is deleted.
```
Comment (Pandrex247) [inline webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResource.java]:
```suggestion
     * If JSON given for the page is null, empty, or an empty JSON object, the page is deleted.
```

File: webapp/src/main/webapp/js/mc-view-components.js [hunk 4]
```diff
@@ -299,31 +299,35 @@ MonitoringConsole.View.Components = (function() {
...
-                  table.append(createRow(entry, createInput(entry)));
-               } else {
-                  if (Array.isArray(input)) {
-                    let [innerInput, innerSyntheticId] = createMultiInput(input, syntheticId, 'x-input');
-                    input = innerInput;
-                    syntheticId = innerSyntheticId;
-                  }
-                  table.append(createRow(entry, input));
-               }
+            if (group.available !== false) {
```
Comment (Pandrex247) [inline webapp/src/main/webapp/js/mc-view-components.js:302]: Double negative?

File: webapp/src/main/webapp/js/mc-view-components.js [hunk 5]
```diff
@@ -1022,6 +1027,98 @@ MonitoringConsole.View.Components = (function() {
...
+        .append($('<td/>', baseAttrs).text(Units.formatDateTime(page.lastRemoteUpdate)))
+        .append($('<td/>', remoteAttrs).text(Units.formatDateTime(page.lastRemoteChange)));
+    }
+
+    return { createComponent: createComponent };
+  })();
+
+
+  /**
+   * A component that creates a panel (for modal dialoge) that presents possible roles for user to select one.
```
Comment (Pandrex247) [inline webapp/src/main/webapp/js/mc-view-components.js]:
```suggestion
   * A component that creates a panel (for modal dialogue) that presents possible roles for user to select one.
```

File: webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResource.java (hunk 2 above)
Comment (jbee) [reply webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResource.java:161]: yes, sadly the API is build in such a way that this is how to read the root element if you assume there is always going to be just one.

File: webapp/src/main/webapp/js/mc-view-components.js (hunk 4 above)
Comment (jbee) [reply webapp/src/main/webapp/js/mc-view-components.js:302]: yes, this is so `true` and `undefined` evaluate to `true` but `false` is still `false`

**Review by Pandrex247 — APPROVED** [review]

**Pull request merged**